source/sigv4.c

```diff
@@ -794,6 +899,332 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+            trimValueLen = 0;
+
+            for( i = 0; i < valLen; i++ )
+            {
+                if( ( isTrimmableSpace( value, i, valLen, trimValueLen ) ) )
+                {
+                    /* Cannot copy trimmable space into canonical request buffer. */
+                }
+                /* Remaining buffer space should at least accommodate the character to copy and the trailing "\n" */
+                else if( ( buffRemaining <= 1 ) )
+                {
+                    sigV4Status = SigV4InsufficientMemory;
+                    break;
+                }
+                else
+                {
+                    *pBufLoc = value[ i ];
+                    pBufLoc++;
+                    trimValueLen++;
+                    buffRemaining -= 1;
+                }
+            }
+
+            if( sigV4Status == SigV4Success )
+            {
+                assert( buffRemaining >= 1 );
+                *pBufLoc = '\n';
+                pBufLoc++;
+
+                /* Calculate the remaining buffer. 1 is added for '\n' character. */
+                buffRemaining -= 1;
+
+                canonicalRequest->pBufCur = pBufLoc;
+                canonicalRequest->bufRemaining = buffRemaining;
+            }
+        }
+
+        return sigV4Status;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t appendCanonicalizedHeaders( size_t headerCount,
+                                                     CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+        assert( headerCount > 0 );
+
+        for( noOfHeaders = 0; noOfHeaders < headerCount; noOfHeaders++ )
+        {
+            assert( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData != NULL );
+            sigV4Status = writeCanonicalHeaderToCanonicalRequest( noOfHeaders, canonicalRequest );
+
+            if( sigV4Status != SigV4Success )
+            {
+                break;
+            }
+        }
+
+        return sigV4Status;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t appendAllHeadersToCanonicalRequest( const char * pHeaders,
+                                                             size_t headersLen,
+                                                             uint32_t flags,
+                                                             CanonicalContext_t * canonicalRequest )
+    {
+        const char * start;
+        const char * end;
+        size_t keyFlag = 1, noOfHeaders = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( pHeaders != NULL );
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        start = end = pHeaders;
```

Comment:
I haven't seen us using double assignment before. Is it acceptable within our coding style?